### DISCUSSION TITLE
Goals for next milestones

### DISCUSSION THREAD

**Discussion by dstebila** [open-quantum-safe/liboqs#1107] [Ideas]
This thread is for discussing what goals we want to work towards in our next release of liboqs as well as other OQS subprojects.

## liboqs

- Algorithm improvements and additions:
  - #1101: Endianness issues with BIKE
  - #1098: Add stateful hash-based signatures
  - #1038: SPHINCS+ memcpy source and destination overlap - undefined behavior
  - #1031: Add NTRU HPS-4096-1229
  - #995: Known non-constant time behaviour in HQC
  - Update Classic McEliece to more recent version
- Code quality
  - #1083: Enabling more compiler warnings
  - #194: scan-build errors
  - #167: Code coverage
- Common code and building
  - #1072: Android Buildscript problem with Cross Compilation
  - #973: Investigate integrating pqcrystals common code to OQS common code
  - #910: Establish interop with Circl
  - #624: Graceful runtime CPU feature handling
  - #594: Add options for iOS and Android cross-compilation

## OpenSSL 

- open-quantum-safe/openssl/issues/333: Add X.509 implementation using composite keys and sigs drafts
- Implement more stuff in oqs-provider
  - open-quantum-safe/oqs-provider/issues/32: Interoperability with oqs-openssl?
  - open-quantum-safe/oqs-provider/issues/17: Hybrid KEM: more combiners, more abstraction
  - open-quantum-safe/oqs-provider/issues/9: Implement full TLS1.3 OQS signature support
  - open-quantum-safe/oqs-provider/issues/8: Implement interoperable CMS
  - open-quantum-safe/oqs-provider/issues/7: Implement interoperable X509 read/write
  - open-quantum-safe/oqs-provider/issues/6: Implement interoperable key gen & storage
  - open-quantum-safe/oqs-provider/issues/3: Add hybrid support
  - open-quantum-safe/oqs-provider/issues/2: Implement encoder/decoder

**Comment by baentsch** [comment]
I can already state that _my_ goal is to complete https://github.com/open-quantum-safe/oqs-provider/issues/2, https://github.com/open-quantum-safe/oqs-provider/issues/3, https://github.com/open-quantum-safe/oqs-provider/issues/7, probably https://github.com/open-quantum-safe/oqs-provider/issues/8 at latest by year-end. It would be nice to have a clear perspective what "Interoperability" shall mean in general (before closing out https://github.com/open-quantum-safe/oqs-provider/issues/32): Across different (TLS) prototype implementations or even implementing IETF drafts? If the latter, which ones?

General question: By when do we want to do a next release? Coinciding with the next NIST decision or before?

**Reply by dstebila** [reply]
> I can already state that _my_ goal is to complete [open-quantum-safe/oqs-provider#2](https://github.com/open-quantum-safe/oqs-provider/issues/2), [open-quantum-safe/oqs-provider#3](https://github.com/open-quantum-safe/oqs-provider/issues/3), [open-quantum-safe/oqs-provider#7](https://github.com/open-quantum-safe/oqs-provider/issues/7), probably [open-quantum-safe/oqs-provider#8](https://github.com/open-quantum-safe/oqs-provider/issues/8) at latest by year-end. It would be nice to have a clear perspective what "Interoperability" shall mean in general (before closing out [open-quantum-safe/oqs-provider#32](https://github.com/open-quantum-safe/oqs-provider/issues/32)): Across different (TLS) prototype implementations or even implementing IETF drafts? If the latter, which ones?

I would say interoperability would be about each of our subprojects implementing the current version of whatever Internet Draft is relevant, and being able to interoperate with each other at the very least at the protocol level.  I don't have strong opinions about whether they should be able to interoperate at a command-line file level, e.g., that private key files saved to disk by OpenSSL 1.1.1 can be used in OpenSSL 3 provider and vice versa.  That would be nice, but in the absence of an external spec saying what format those key files should have, I don't see it as strictly necessary.

> General question: By when do we want to do a next release? Coinciding with the next NIST decision or before?

How about a release by the end of November?  That would more or less tie a ribbon on everything pre-NIST decision and then we can decide how to move forward once we see the NIST decision.

**Comment by baentsch** [comment]
Agree to all of the above with one caveat:

>  That would be nice, but in the absence of an external spec saying what format those key files should have, I don't see it as strictly necessary.

In the absence of an external spec I'd see it as _very sensible_ to not create more different file formats, particularly by a single team. It's bad enough when/if different companies create non-interoperable QSC data structures but an open source project should strive to avoid that/do that only with a good reason (say implementing a draft RFC).

In addition we'll have more/stronger NTRUPrime algs (#1111) in an 0.7.1 release.

**Comment by christianpaquin** [comment]
Just added issue #1113 (new Picnic update); should be quick and one more thing for 0.7.1
